src/rnnlm/rnnlm-lattice-rescoring.cc

```diff
@@ -0,0 +1,169 @@
+    fst_label_to_word_[i] = fst_word_symbols->Find(i);
+    if (fst_label_to_word_[i] == "") {
+      KALDI_ERR << "Could not find word for integer " << i << "in the word "
+                << "symbol table, mismatched symbol table or you have discoutinuous "
+                << "integers in your symbol table?";
+    }
+  }
+
+//  fst_label_to_rnn_out_label_.resize(fst_word_symbols->NumSymbols(), -1);
+  KALDI_LOG << "resize to " << fst_word_symbols->NumSymbols();
+  fst_label_to_rnn_label_.resize(fst_word_symbols->NumSymbols(), -1);
+
+  out_OOS_index_ = 1;
+  {
+    std::ifstream ifile(rnn_wordlist.c_str());
+    int32 id;
+    string word;
+    int32 i = 0;
+    while (ifile >> word >> id) {
+      if (word == "</s>") {
+        final_word_index_ = id;
+      }
+      KALDI_ASSERT(i == id);
+      i++;
+      rnn_label_to_word_.push_back(word);
+
+//      if (word == "<brk>") {
+//        continue;
+//      }
+
+      int fst_label = fst_word_symbols->Find(rnn_label_to_word_[id]);
+      if (fst::SymbolTable::kNoSymbol != fst_label || id == out_OOS_index_ || id == 0 || word == "<brk>") {}
+      else {
+        KALDI_LOG << "warning: word " << word << " in RNNLM wordlist but not in FST wordlist";
+      }
+      if (id != out_OOS_index_ && out_OOS_index_ != 0 && fst_label != -1) {
+        fst_label_to_rnn_label_[fst_label] = id;
+      }
+    }
+  }
+
+  for (int32 i = 0; i < fst_label_to_rnn_label_.size(); i++) {
+    if (fst_label_to_rnn_label_[i] == -1) {
+      fst_label_to_rnn_label_[i] = out_OOS_index_;
+    }
+  }
+}
+
+KaldiRnnlmDeterministicFst::KaldiRnnlmDeterministicFst(int32 max_ngram_order,
+    const std::string &rnn_wordlist,
+    const std::string &word_symbol_table_rxfilename,
+    const DecodableRnnlmSimpleLoopedInfo &info) {
+  max_ngram_order_ = max_ngram_order;
+  ReadFstWordSymbolTableAndRnnWordlist(rnn_wordlist,
+                                       word_symbol_table_rxfilename);
+
+  std::vector<Label> bos;
+  bos.push_back(0); // 0 for <s>
+  state_to_wseq_.push_back(bos);
+  DecodableRnnlmSimpleLooped decodable_rnnlm(info);
+  decodable_rnnlm.TakeFeatures(std::vector<Label>(1, bos[0]));
+  state_to_decodable_rnnlm_.push_back(decodable_rnnlm);
+  wseq_to_state_[bos] = 0;
+  start_state_ = 0;
+}
+
+fst::StdArc::Weight KaldiRnnlmDeterministicFst::Final(StateId s) {
+  // At this point, we should have created the state.
+  KALDI_ASSERT(static_cast<size_t>(s) < state_to_wseq_.size());
+
+  // log prob of end of sentence
+  BaseFloat logprob = state_to_decodable_rnnlm_[s].GetOutput(0, final_word_index_);
+  return Weight(-logprob);
+}
+
+bool KaldiRnnlmDeterministicFst::GetArc(StateId s, Label ilabel,
+                                        fst::StdArc *oarc) {
+  // At this point, we should have created the state.
+  KALDI_ASSERT(static_cast<size_t>(s) < state_to_wseq_.size());
+
+  std::vector<Label> wseq = state_to_wseq_[s];
+  DecodableRnnlmSimpleLooped decodable_rnnlm = state_to_decodable_rnnlm_[s];
```

Comment:
It looks like you are implicitly using the copy constructor of this class.  You should make this explicit by writing out the copy constructor (and also make the assignment operator private to disallow its use).

However, there are some problems with how this is written.  When the storage of state_to_decodable_rnnlm_  is reallocated, expensive copies and destructions of the DecodableRnnlmSimpleLooped objects will take place.   state_to_decodable_rnnlm_[s] should instead contain pointers.  And you should allocate these objects via "new", instead of on the stack.  For non-initial states, your "new" operation can use the self-constructor.  The issue with allocating this on the stack  is that you have to copy and then destroy this object, which is very wasteful.